### DISCUSSION TITLE
提高协程帧的分配性能的思路

### DISCUSSION THREAD

**Discussion by microcai** [alibaba/async_simple#398] [Ideas]
为了提高协程的性能，可以使用 一种叫栈试式内存分配器
比如 A 调用 B。肯定是 A 先分配。
B 后分配。
释放的时候，肯定也是 B 先释放。
然后 A 才被释放。
对吧？
既然如此，在自定义协程帧的分配器的时候，使用栈式分配策略。先一次性向 os 申请一片连续的内存。然后，每单创建一个新协程promise，就移动一下指针完成分配。释放的时候，也是简单的回退指针完成分配。

只有处在同一条调用链上的一系列协程函数，才共享同一个分配器。
也就是说，这个分配器最好是使用“协程本地存储”的方式存起来。然后每个协程函数，都通过获取协程本地存储的方式，获得这个分配器。再用这个分配器去分配新的 await promise 对象。

**Comment by ChuanqiXu9** [comment]
> 比如 A 调用 B。肯定是 A 先分配。
B 后分配。
释放的时候，肯定也是 B 先释放。
然后 A 才被释放。
对吧？

这个不一定的。哪怕是 lazy，也只是在 `co_await` 时才会有这个效果。可以看下这个：https://clang.llvm.org/docs/AttributeReference.html#coro-await-elidable

**Reply by microcai** [reply]
不是在栈上分配。而是使用 栈式内存分配策略的内存分配器分配 协程帧。
当然前提是协程库 支持自定义 分配器

**Reply by ChuanqiXu9** [reply]
明白了。现在新增了 https://github.com/alibaba/async_simple/pull/393 。可能可以做。但接口可能会比较丑

**Reply by microcai** [reply]
重载 operator new 。 在 operator new 里调用传入的迭代器. 每个需要使用迭代器分配的 协程函数，都要声明带上自己并不使用的 迭代器参数。
那这样的做法，和 asio 的 stackfull 协程里，每个协程都得带上 yield_context 有什么区别了？
这种设计过于丑陋。有没有办法从 协程本地存储里直接获取 分配器进行内存分配。这样只要  lauched_detached 的时候，设置一次分配器就可以了。

**Reply by ChuanqiXu9** [reply]
我目前能想到的办法只能是让编译器配合了，按我理解现存的语言设计不支持这么做。

**Reply by 20083017** [reply]
> 重载 operator new 。 在 operator new 里调用传入的迭代器. 每个需要使用迭代器分配的 协程函数，都要声明带上自己并不使用的 迭代器参数。 那这样的做法，和 asio 的 stackfull 协程里，每个协程都得带上 yield_context 有什么区别了？ 这种设计过于丑陋。有没有办法从 协程本地存储里直接获取 分配器进行内存分配。这样只要 lauched_detached 的时候，设置一次分配器就可以了。

看着seastar里边也有自定义allocator的内容，好像只是针对native-stack生效，然后每个reactor初始化的时候都做了对齐啥的，这块儿对性能的形象有多大呀 大佬给解释下呗 (^_^)

**Reply by ChuanqiXu9** [reply]
不是对齐，是动态内存分配对性能的影响，可以搜索 “动态分配对性能的多少影响”，有很多已有资料。
